Open graph/metadata improvements (initial concept)

inxilpro opened the pull request:
Every time the Laravel docs are posted on Twitter and a fully generic open graph card pops up, I say to myself, "someday I'm going to PR a fix for that." So, this is my proposed fix:

- Allow for Markdown "front matter" in docs files
- If that metadata exists, use that in the open graph and Twitter `<meta>` tags
- If not, use the current defaults

I extracted out a `Page` class to encapsulate some of that logic, and had to tweak the layout a bit to allow for overriding the metadata. I've included one sample open graph image—should this PR move forward, I will provide a better mechanism for the open graph images (I just mocked this up in Photoshop for now, but I hope there's a way to use Torchlight and something like Browsershot to generate them).

<table>
<tr>
<td>
<img src="https://user-images.githubusercontent.com/21592/195954653-f5d79302-b70c-4d58-9edf-6c4f829ecf3b.png" alt="Before" />
</td>
<td>
<img src="https://user-images.githubusercontent.com/21592/195954870-584fdcc9-62b1-4a30-970a-c72f845f41b5.png" alt="After" />
</td>
</tr>
</table>

This is the front matter that would achieve this:

```md
---
description: Get started with Eloquent, Laravel's object-relational mapper (ORM) that makes it enjoyable to interact with your database.
og:
  image: og/eloquent.jpg
---

# Eloquent ORM

...
```

Should this PR be merged, the next step would be updating all the markdown files in the [docs repo](https://github.com/laravel/docs) to include necessary front matter. At the same time, we can add open graph images, or do that in a second pass.

driesvints commented:
I'm not sure we actually want to commit the images? Can we ignore them and generate them on deploy?

inxilpro commented:
> I'm not sure we actually want to commit the images? Can we ignore them and generate them on deploy?

I would be OK with generating them on deploy. I thought this might be an easier starting point (since all the other images are committed to the repo, and generating them does come with some overhead). Realistically, we could commit this without any of the images now and then worry about that later, since the docs markdown files will all need to be updated before the images can be generated.

inxilpro commented:
@driesvints I just updated the PR:

- Added a sample workflow for auto-generating open graph images via github actions. I'm not 100% sure how Laravel.com is deployed, so I'm not certain this is the right approach
- Removed the sample images from this PR
- Updated the code to use the first code block if a custom one isn't present

To generate open graph images, you need to:

```shell
chromedriver --port=4444 &
php artisan serve &
php artisan og:generate-images
```

This will push a bunch of images to `public/img/og/`

Let me know if you'd like me to do anything else to help push this over the finish line.

inxilpro commented:
Oh interesting. Once I got the github workflow working it auto-committed the images to the repo. Let me know what you think makes the most sense, and I can update accordingly.

driesvints commented:
Hah I have to admit that workflow is pretty neat. The only thing is that it seems to take a while (4 minutes). If you push a lot it would trigger quite a few long running builds. I think it's best that if we go forward with generating those images we only build on master branch. But honestly I still think we should maybe set up a scheduled job that runs once a day to generate these. Committing all of these to the repo would increase its size unnecessarily.

inxilpro commented:
@driesvints how is Laravel.com deployed? We probably don't need to even run it every day… we could just run it as a post-deploy script. The only issue with running the command on the Laravel.com server is making sure that it's set up to run chromedriver/etc.

Two other options:

1. Push the images somewhere like S3 rather than into the git repo
2. Pull this code into a separate repository and then pull the images from that repo on deploy

driesvints commented:
Through Forge. I don't think those two options make sense imo, sorry. Think it's best to either do a scheduled task or on deploy. 

@taylorotwell what are your thoughts on this PR?

inxilpro commented:
To summarize, here are the two options that seem to make the most sense:

1. Go with the PR as-is, where images are generated via Github Actions and committed directly to the repo. It's possible to mitigate the overhead here by both a) only generating on push to `master`, and b) saving a manifest file that only rebuilds images when the source docs have changed.
2. Run `php artisan og:generate-images` on a schedule on the production server. This way, the images don't need to be committed to the repo. The downside is that then `chromedriver` needs to be installed and kept up-to-date on the production server, and the existing command would need to be updated to spawn the `chromedriver` instance in the background (much like Dusk).

I think both options are fine. Option 1 is much simpler, but if repository size is a concern, I'm willing to make option 2 work.

inxilpro commented:
@driesvints just FYI—by applying some optimizations to the workflow, I was able to get the run down to [about a minute and a half](https://github.com/inxilpro/laravel.com-fork/actions/runs/3339483703/jobs/5528488276). It no longer downloads every version of the docs, caches composer data, and only generates images for markdown files that have actually changed since the last run.

inxilpro commented:
I see that @taylorotwell marked this as draft — let me know if there's anything you need from me on this.

TheBlckbird commented:
Is this still being worked on?

inxilpro commented:
> Is this still being worked on?

I'm just waiting on @driesvints or @taylorotwell to make a call. If there's renewed interest in this PR, I'm happy to make changes or get it caught up with the current codebase…

shaedrich commented:
That would be quite nice to have
